File: chap1/sec1.5/prob1.1/sol_err.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu Sep 13 15:35:52 2018

@author: lunet
"""
import numpy as np
import lorenz as lo


# Definition of the function used for time integration
def f(t, u):
    return lo.lorenzOperator(u[0], u[1], u[2])


# Local error computation
nStep = 1
nStepRef = 1000
dt = np.array([0.0001, 0.001, 0.01, 0.1])
lErrLoc = []
for T in dt:
    t, u = lo.forwardEuler(f, 0, T, [20, 5, -5], nStep)
    t, uRef = lo.forwardEuler(f, 0, T, [20, 5, -5], nStepRef)
    err = np.linalg.norm(uRef[:, -1] - u[:, -1])
    lErrLoc.append(err)

# Global error computation
T = max(dt)
lStep = np.array(T/dt, dtype=int)
nStepRef = 100*max(lStep)
t, uRef = lo.forwardEuler(f, 0, T, [20, 5, -5], nStepRef)
lErrGlob = []
for nStep in lStep:
    t, u = lo.forwardEuler(f, 0, T, [20, 5, -5], nStep)
    err = np.linalg.norm(uRef[:, -1] - u[:, -1])
    lErrGlob.append(err)

# Q-1(h): error plot
lo.plot2DCurve(dt, lErrLoc, 'Err', 'o--', label='Loc. trun. error')
lo.plot2DCurve(dt, lErrGlob, 'Err', 's--', label='Glob. trun. error')
lo.plot2DCurve(dt, 10**3.5 * dt, 'Err', label='Order 1')
lo.plot2DCurve(dt, 10**3 * dt**2, 'Err', label='Order 2',
               logX=True, logY=True, xLabel='$\\Delta_t$')

lo.plt.show()

File: chap1/sec1.5/prob1.1/sol_err.m
% Settings for Matlab
close all; clear;
set( 0, 'defaultAxesTickLabelInterpreter', 'latex' );
set( 0, 'defaultLegendInterpreter',        'latex' );
set( 0, 'defaultTextInterpreter',          'latex' );
set( 0, 'defaultAxesFontSize', 12 );
format long 

% Definition of the function used for time integration
sigma = 10; rho=28; beta = 8/3;
f=@(t,u) LorenzOperator(u(1), u(2), u(3), sigma, rho, beta);

% Local error computation
nStep = 1;
nStepRef = 1000;
dt = [0.0001, 0.001, 0.01, 0.1];
lErrLoc = [];
for T = dt
    [~, u] = ForwardEuler(f, 0, T, [20, 5, -5], nStep);
    [~, uRef] = ForwardEuler(f, 0, T, [20, 5, -5], nStepRef);
    err = norm(uRef(:, end) - u(:, end));
    lErrLoc = [lErrLoc, err];
end

% Global error computation
T = max(dt);
lStep = T./dt;
nStepRef = 100*max(lStep);
[t, uRef] = ForwardEuler(f, 0, T, [20, 5, -5], nStepRef);
lErrGlob = [];
for nStep = lStep
    [~, u] = ForwardEuler(f, 0, T, [20, 5, -5], nStep);
    err = norm(uRef(:, end) - u(:, end));
    lErrGlob = [lErrGlob, err];
end

% Q-1(h): error plot
loglog(dt, lErrLoc, '--o', 'DisplayName', 'Loc. trun. error');
hold on;
loglog(dt, lErrGlob, '--s', 'DisplayName', 'Glob. trun. error');
loglog(dt, 10^3.5 * dt, 'DisplayName', 'Order 1');
loglog(dt, 10^3 * dt.^2, 'DisplayName', 'Order 1');
xlabel('$\Delta t$'); legend('Location', 'northwest'); grid on;
